Get rid of some `#![allow(static_mut_refs)]`

GrigorenkoPV opened the pull request.

oli-obk commented:
r=me on the compiler changes

r? @RalfJung on the miri test changes. They lgtm, but maybe there are subtle interactions that should have been handled differently

RalfJung commented:
Miri changes LGTM, thanks!
@bors r=oli-obk,RalfJung

bors commented:
:pushpin: Commit 859c1225ce339a46ec795d5b138a0d181222b549 has been approved by `oli-obk,RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk,RalfJung 859c1225ce339a46ec795d5b138a0d181222b549 -->
<!-- homu: {"type":"Approved","sha":"859c1225ce339a46ec795d5b138a0d181222b549","approver":"oli-obk,RalfJung","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 859c1225ce339a46ec795d5b138a0d181222b549 with merge daa42644a51931b6e6b5d799040d54e6647f7eef...
<!-- homu: {"type":"BuildStarted","head_sha":"859c1225ce339a46ec795d5b138a0d181222b549","merge_sha":"daa42644a51931b6e6b5d799040d54e6647f7eef"} -->

rust-log-analyzer commented:
The job **`dist-aarch64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/8014395382/job/21892919999) [(plain)](https://github.com/rust-lang-ci/rust/commit/daa42644a51931b6e6b5d799040d54e6647f7eef/checks/21892919999/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling panic_abort v0.0.0 (C:\a\rust\rust\library\panic_abort)
   Compiling std_detect v0.1.5 (C:\a\rust\rust\library\stdarch\crates\std_detect)
   Compiling hashbrown v0.14.3
[RUSTC-TIMING] rustc_std_workspace_alloc test:false 0.040
error: shared reference of mutable static is discouraged
   --> library\panic_unwind\src\seh.rs:274:63
    |
274 |     let mut exception = ManuallyDrop::new(Exception { canary: &TYPE_DESCRIPTOR, data: Some(data) });
    |
    = note: for more information, see issue #114447 <https://github.com/rust-lang/rust/issues/114447>
    = note: reference of mutable static is a hard error from 2024 edition
    = note: mutable statics can be written to by multiple threads: aliasing violations or data races will cause undefined behavior
    = note: mutable statics can be written to by multiple threads: aliasing violations or data races will cause undefined behavior
    = note: `-D static-mut-ref` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(static_mut_ref)]`
help: shared references are dangerous since if there's any kind of mutation of that static while the reference lives, that's UB; use `addr_of!` instead to create a raw pointer
    |
274 |     let mut exception = ManuallyDrop::new(Exception { canary: addr_of!(TYPE_DESCRIPTOR), data: Some(data) });

error: shared reference of mutable static is discouraged
error: shared reference of mutable static is discouraged
   --> library\panic_unwind\src\seh.rs:303:14
    |
303 |         ptr!(&CATCHABLE_TYPE_ARRAY as *const _) as u32,
    |
    = note: for more information, see issue #114447 <https://github.com/rust-lang/rust/issues/114447>
    = note: reference of mutable static is a hard error from 2024 edition
    = note: mutable statics can be written to by multiple threads: aliasing violations or data races will cause undefined behavior
    = note: mutable statics can be written to by multiple threads: aliasing violations or data races will cause undefined behavior
help: shared references are dangerous since if there's any kind of mutation of that static while the reference lives, that's UB; use `addr_of!` instead to create a raw pointer
    |
303 |         ptr!(addr_of!(CATCHABLE_TYPE_ARRAY) as *const _) as u32,

error: shared reference of mutable static is discouraged
error: shared reference of mutable static is discouraged
   --> library\panic_unwind\src\seh.rs:307:14
    |
307 |         ptr!(&CATCHABLE_TYPE as *const _) as u32,
    |
    = note: for more information, see issue #114447 <https://github.com/rust-lang/rust/issues/114447>
    = note: reference of mutable static is a hard error from 2024 edition
    = note: mutable statics can be written to by multiple threads: aliasing violations or data races will cause undefined behavior
    = note: mutable statics can be written to by multiple threads: aliasing violations or data races will cause undefined behavior
help: shared references are dangerous since if there's any kind of mutation of that static while the reference lives, that's UB; use `addr_of!` instead to create a raw pointer
    |
307 |         ptr!(addr_of!(CATCHABLE_TYPE) as *const _) as u32,

error: shared reference of mutable static is discouraged
error: shared reference of mutable static is discouraged
   --> library\panic_unwind\src\seh.rs:311:14
    |
311 |         ptr!(&TYPE_DESCRIPTOR as *const _) as u32,
    |
    = note: for more information, see issue #114447 <https://github.com/rust-lang/rust/issues/114447>
    = note: reference of mutable static is a hard error from 2024 edition
    = note: mutable statics can be written to by multiple threads: aliasing violations or data races will cause undefined behavior
    = note: mutable statics can be written to by multiple threads: aliasing violations or data races will cause undefined behavior
help: shared references are dangerous since if there's any kind of mutation of that static while the reference lives, that's UB; use `addr_of!` instead to create a raw pointer
    |
311 |         ptr!(addr_of!(TYPE_DESCRIPTOR) as *const _) as u32,

error: mutable reference of mutable static is discouraged
error: mutable reference of mutable static is discouraged
   --> library\panic_unwind\src\seh.rs:322:35
    |
322 |     _CxxThrowException(throw_ptr, &mut THROW_INFO as *mut _ as *mut _);
    |
    = note: for more information, see issue #114447 <https://github.com/rust-lang/rust/issues/114447>
    = note: reference of mutable static is a hard error from 2024 edition
    = note: mutable statics can be written to by multiple threads: aliasing violations or data races will cause undefined behavior
    = note: mutable statics can be written to by multiple threads: aliasing violations or data races will cause undefined behavior
help: mutable references are dangerous since if there's any other pointer or reference used for that static while the reference lives, that's UB; use `addr_of_mut!` instead to create a raw pointer
    |
322 |     _CxxThrowException(throw_ptr, addr_of_mut!(THROW_INFO) as *mut _ as *mut _);

error: shared reference of mutable static is discouraged
error: shared reference of mutable static is discouraged
   --> library\panic_unwind\src\seh.rs:333:25
    |
333 |     if !ptr::eq(canary, &TYPE_DESCRIPTOR) {
    |
    = note: for more information, see issue #114447 <https://github.com/rust-lang/rust/issues/114447>
    = note: reference of mutable static is a hard error from 2024 edition
    = note: mutable statics can be written to by multiple threads: aliasing violations or data races will cause undefined behavior
    = note: mutable statics can be written to by multiple threads: aliasing violations or data races will cause undefined behavior
help: shared references are dangerous since if there's any kind of mutation of that static while the reference lives, that's UB; use `addr_of!` instead to create a raw pointer
    |
333 |     if !ptr::eq(canary, addr_of!(TYPE_DESCRIPTOR)) {

[RUSTC-TIMING] panic_abort test:false 0.053
[RUSTC-TIMING] panic_unwind test:false 0.074
error: could not compile `panic_unwind` (lib) due to 6 previous errors
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8014395382/job/21892998522)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/8014395382/job/21892998522","builder_name":"checks-actions"} -->

GrigorenkoPV committed:
Get rid of some `#[allow(static_mut_refs)]`

GrigorenkoPV commented:
Hopefully this is fixed now.

I have also rebased on top of master and squashed all the changes in one commit.

Will probably look into the ways of setting up a win-msvc toolchain on linux, with some terrible wine hacking, because the experience I have just had with a Windows VM was not very pleasant. But at least more responsive than a CI failure.

RalfJung commented:
The stuff @saethlin did for mir-opt tests should let us do check-builds for MSVC targets on a Linux host, without any Windows tools around.

GrigorenkoPV commented on library/panic_unwind/src/seh.rs:
```diff
@@ -220,7 +220,7 @@ extern "C" {
 // This is fine since the MSVC runtime uses string comparison on the type name
 // to match TypeDescriptors rather than pointer equality.
 static mut TYPE_DESCRIPTOR: _TypeDescriptor = _TypeDescriptor {
-    pVFTable: unsafe { &TYPE_INFO_VTABLE } as *const _ as *const _,
```
Yes, it said something about external statics iirc

GrigorenkoPV replied:
> Strange, did this one not trigger the warning? It doesn't seem to be inside any of the `allow`.

Ah, I guess this is because it is not a `mut` static:
https://github.com/rust-lang/rust/pull/121303/files#diff-1308e363d41cbb1f3f93d560e000263610240cc456596918347038d303dac7b6L214

RalfJung replied:
Ah I see, you just made them all use addr_of for good measure. Makes sense!

GrigorenkoPV replied:
> Ah I see, you just made them all use addr_of for good measure. Makes sense!

I think there are many more places where `& ... as *const ...` is used. Would it make sense to implement a lint (probably allow-by-default) suggesting to replace those with `addr_of!`? Or maybe at least for me to go through the libs and change those manually?

RalfJung replied:
Yeah those should all be addr_of ideally.

RalfJung commented on library/std/src/panicking.rs:
```diff
@@ -337,9 +337,6 @@ pub mod panic_count {
 #[unstable(feature = "update_panic_count", issue = "none")]
-// FIXME: Use `SyncUnsafeCell` instead of allowing `static_mut_refs` lint
-#[cfg_attr(bootstrap, allow(static_mut_ref))]
-#[cfg_attr(not(bootstrap), allow(static_mut_refs))]
```
Looks like no fix was needed for these. This does make me wonder why they were ever added... :shrug:

RalfJung commented:
@bors r=oli-obk,RalfJung rollup=iffy

bors commented:
:pushpin: Commit 58c8c0853f08d83a86edfc9cd73463ff29f0509f has been approved by `oli-obk,RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk,RalfJung 58c8c0853f08d83a86edfc9cd73463ff29f0509f -->
<!-- homu: {"type":"Approved","sha":"58c8c0853f08d83a86edfc9cd73463ff29f0509f","approver":"oli-obk,RalfJung","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

saethlin commented:
> The stuff @saethlin did for mir-opt tests should let us do check-builds for MSVC targets on a Linux host, without any Windows tools around.

To be clear, `x check library --target=aarch64-pc-windows-msvc` _just works_. You should be able to set various targets for `x check` without worrying about what host you're on, though this is of course not tested in all the various ways that bootstrap can be used.

RalfJung commented:
Ah, great. :)

We should add that to PR CI for some targets... currently e.g. macOS is only covered by the Miri tests I think.

bors commented:
:hourglass: Testing commit 58c8c0853f08d83a86edfc9cd73463ff29f0509f with merge 2dbd6233ccdb2cd4b621a5e839a95c3fbbc0c375...
<!-- homu: {"type":"BuildStarted","head_sha":"58c8c0853f08d83a86edfc9cd73463ff29f0509f","merge_sha":"2dbd6233ccdb2cd4b621a5e839a95c3fbbc0c375"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8023735261/job/21924514097)
Approved by: oli-obk,RalfJung
Pushing 2dbd6233ccdb2cd4b621a5e839a95c3fbbc0c375 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk,RalfJung","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8023735261/job/21924514097"},"merge_sha":"2dbd6233ccdb2cd4b621a5e839a95c3fbbc0c375"} -->

rust-timer commented:
Finished benchmarking commit (2dbd6233ccdb2cd4b621a5e839a95c3fbbc0c375): [comparison URL](https://perf.rust-lang.org/compare.html?start=21033f637e60af61ead04cc296e7214f89c16a95&end=2dbd6233ccdb2cd4b621a5e839a95c3fbbc0c375&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=21033f637e60af61ead04cc296e7214f89c16a95&end=2dbd6233ccdb2cd4b621a5e839a95c3fbbc0c375&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=21033f637e60af61ead04cc296e7214f89c16a95&end=2dbd6233ccdb2cd4b621a5e839a95c3fbbc0c375&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.4% | [-3.4%, -3.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -3.4% | [-3.4%, -3.4%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=21033f637e60af61ead04cc296e7214f89c16a95&end=2dbd6233ccdb2cd4b621a5e839a95c3fbbc0c375&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 5.3% | [5.2%, 5.5%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=21033f637e60af61ead04cc296e7214f89c16a95&end=2dbd6233ccdb2cd4b621a5e839a95c3fbbc0c375&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 650.563s -> 650.231s (-0.05%)
**Artifact size**: 310.99 MiB -> 310.96 MiB (-0.01%)
